### DISCUSSION TITLE
Discovered incompatible solidity versions

### DISCUSSION THREAD

**Discussion by AnouarBF** [Cyfrin/foundry-full-course-cu#667] [Q&A] [answered]
I have never got this problem before. I tried ```forge build``` and the compiler just throws this error message : 
```
[⠊] Compiling...2023-08-19T07:31:36.130966Z ERROR ethers_solc::resolver: failed to resolve versions
[⠒] Compiling...
Error: 
Discovered incompatible solidity versions in following
: src/DSCEngine.sol (^0.8.18) imports:
    src/StableCoin.sol (^0.8.18)
    lib/openzeppelin-contracts/contracts/security/ReentrancyGuard.sol (^0.8.0)
    lib/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol (^0.8.0)
    lib/chainlink-brownie-contracts/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol (^0.6.0)
    lib/openzeppelin-contracts/contracts/token/ERC20/extensions/ERC20Burnable.sol (^0.8.0)
    lib/openzeppelin-contracts/contracts/access/Ownable2Step.sol (^0.8.0)
    lib/openzeppelin-contracts/contracts/token/ERC20/ERC20.sol (^0.8.0)
    lib/openzeppelin-contracts/contracts/utils/Context.sol (^0.8.0)
    lib/openzeppelin-contracts/contracts/access/Ownable.sol (^0.8.0)
    lib/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol (^0.8.0)
    lib/openzeppelin-contracts/contracts/token/ERC20/extensions/IERC20Metadata.sol (^0.8.0)
    lib/openzeppelin-contracts/contracts/utils/Context.sol (^0.8.0)
    lib/openzeppelin-contracts/contracts/utils/Context.sol (^0.8.0)
    lib/openzeppelin-contracts/contracts/token/ERC20/IERC20.sol (^0.8.0)
```
I tried to reinstall all dependencies and got the same problem.

**Comment by theirrationalone** [comment]
if you're on windows and using wsl then do this....

Go to this directory: `\\wsl.localhost\Ubuntu\home\theirrationalone\.svm` after reaching on .svm folder -> delete everything from it and then try again....

**Reply by AnouarBF** [reply]
No I'm using linux ubuntu

**Reply by theirrationalone** [reply]
try this

add this into your foundry.toml

```toml
solc = "^0.8.0"

```

**Reply by AnouarBF** [reply]
Now it gives me this 
```
Error: 
`solc` ^0.8.0 does not exist
```
I think I need to install "solc" but how ?

**Reply by theirrationalone** [reply]
now run foundryup

**Reply by AnouarBF** [reply]
still gives me this : 
```
Error: 
`solc` ^0.8.0 does not exist
```

**Reply by theirrationalone** [reply]
try with this one...

```toml
solc = "0.8.0"
```

**Reply by AnouarBF** [reply]
The problem here is I have different solidity versions between chainlink(0.6.0) and opezeppelin(0.8.0)

**Reply by theirrationalone** [reply]
Foundry does this for us automatically, But In your case, Somehow it's not working.... One mitigation we can do...

Try to reinstall `openzeppelin-contracts@v4.8.3` like this...

```terminal
forge install openzeppelin/openzeppelin-contracts@v4.8.3 --no-commit
```

**Reply by AnouarBF** [reply]
Still the same! I've done this already and didn't work

**Reply by AnouarBF** [reply]
didn't thought it will be that complicated problem :\

**Reply by theirrationalone** [reply]
Now Only last thing we can do is, Please find Foundry's root and delete everything that's into the .cache folder(Don't delete .cache folder itself)...

Foundry's root structure might appears like this...

![Screenshot (105)](https://github.com/Cyfrin/foundry-full-course-f23/assets/96755510/5603f727-d9ac-40ca-bb38-cb00db70df20)

**Comment by alfheimrShiven** [comment, marked as answer]
@AnouarBF you seem to be using `v0.6` of `AggregatorV3Interface`. This uses solidity version `^0.6.0` while our `DSCEngine.sol` uses `^0.8.xx` which are incompatible. Instead, use the `v0.8` of `AggregatorV3Interface` in your `/src/DSCEngine.sol`:

`import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.6/interfaces/AggregatorV3Interface.sol"; ❌`
`import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol"; ✅`

Now run `forge build`. It should compile without any errors.

**Reply by AnouarBF** [reply]
Oh yees How didn't I noticed that. I've had to rebuild all the project and finished it. Now I know where is the problem. Thank you so much @alfheimrShiven

**Reply by alfheimrShiven** [reply]
Cheers @AnouarBF . Glad I could help.
